examples/federated-learning/tff/nvflare/README.md

```diff
@@ -0,0 +1,313 @@
+
+- A persistent volume to store the NVFlare workspace
+- Two pods that are the clients that will be connected to the server in the nvidia-client1 and nvidia-client2 namespaces respectively
+- One pod that is the the server that will aggregate all the results from the computation in the nvflare-infra namespace
+
+## Deploy the blueprint
+
+This example builds on top of the infrastructure that the
+[blueprint provides](../../../../README.md), and follows the best practices the
+blueprint establishes.
+
+To deploy the NVFlare demo described in this document, you need to deploy the [Federated learning blueprint](../../../../README.md#deploy-the-blueprint) first. Then, you can deploy the NVFlare demo described in this document.
+
+### Build and push the example container image
+
+You have to create a custom image with TensorFlow and NVFlare installed. From the main `terraform` folder:
+
+1. Open [Cloud Shell](https://cloud.google.com/shell)
+
+1. Build the example container image locally on your host:
+
+   ```bash
+   NVFLARE_EXAMPLE_CONTAINER_IMAGE_REPOSITORY_HOSTNAME="$(terraform output -raw container_image_repository_fully_qualified_hostname)"
+   NVFLARE_EXAMPLE_CONTAINER_IMAGE_LOCALIZED_ID="${NVFLARE_EXAMPLE_CONTAINER_IMAGE_REPOSITORY_HOSTNAME}/$(terraform output -raw container_image_repository_name)/nvflare-tensorflow"
+   NVFLARE_EXAMPLE_CONTAINER_IMAGE_BUILD_CONTEXT_PATH="../examples/federated-learning/tff/nvflare/container-image"
+   export NVFLARE_EXAMPLE_CONTAINER_IMAGE_TAG="0.0.1"
+   NVFLARE_EXAMPLE_CONTAINER_IMAGE_LOCALIZED_ID_WITH_TAG=${NVFLARE_EXAMPLE_CONTAINER_IMAGE_LOCALIZED_ID}:${NVFLARE_EXAMPLE_CONTAINER_IMAGE_TAG}
+   export NVFLARE_EXAMPLE_CONTAINER_IMAGE_LOCALIZED_ID
+
+   docker build \
+     --file "${NVFLARE_EXAMPLE_CONTAINER_IMAGE_BUILD_CONTEXT_PATH}/Dockerfile" \
+     --tag "${NVFLARE_EXAMPLE_CONTAINER_IMAGE_LOCALIZED_ID_WITH_TAG}" \
+     ${NVFLARE_EXAMPLE_CONTAINER_IMAGE_BUILD_CONTEXT_PATH}
+   ```
+
+   It takes about 8GBs of persistant storage space and can take up to 20 minutes to build. For a production deployment, consider using Cloud Build.
+
+1. Authenticate Docker with the Artifact Registry repository:
+
+   ```bash
+   gcloud auth configure-docker \
+       "${NVFLARE_EXAMPLE_CONTAINER_IMAGE_REPOSITORY_HOSTNAME}"
+   ```
+
+1. Push the container image to the Artifact Registry repository:
+
+   ```bash
+   docker image push "${NVFLARE_EXAMPLE_CONTAINER_IMAGE_LOCALIZED_ID_WITH_TAG}"
+   ```
+
+### Create the infrastructure to store the models
+
+All the models generated will be stored in a Cloud storage bucket mounted by each pod.
+
+1. Initialize the following Terraform variables:
+
+   ```hcl
+   nvflare               = true
+   workspace_bucket_name = # Bucket name that will contain nvflare workspace
+   ```
+
+   The bucket name has to be unique. To generate a unique name for your bucket, you can use this command
```

Comment:
Move `To generate a unique name for your bucket, you can use this command` over a new line, and refactor it as:

For example, you can generate a unique name for your bucket with the following command:

```
RANDOM_SUFFIX=$(echo $RANDOM | md5sum | head -c 4; echo;)
echo "nvflare-storage-${RANDOM_SUFFIX}"
```

Reply:
Done